### PR TITLE
PYTHON-3336 Test Failure - test_load_balancer failing

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: pymongo/errors.py [hunk 1]
```diff
@@ -61,6 +61,15 @@ class ConnectionFailure(PyMongoError):
     """Raised when a connection to the database cannot be made or is lost."""
 
 
+class WaitQueueTimeoutError(ConnectionFailure):
+    """Raised when we time out due to the query wait time.
```
Comment (ShaneHarvey) [inline pymongo/errors.py:65]: "Raised when an operation times out waiting to checkout a connection from the pool."

File: pymongo/mongo_client.py [hunk 2]
```diff
@@ -2111,14 +2111,16 @@ def handle(self, exc_type, exc_val):
                 ):
                     self.session._unpin()
 
+        completed_handshake = self.server.pool.handshake
```
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/mongo_client.py:2114]: Ack, I'll keep digging

File: pymongo/errors.py (hunk 1 above)
Comment (blink1073) [reply pymongo/errors.py:65]: Fixed

File: pymongo/mongo_client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/mongo_client.py]: Is `self.server.pool.state == PoolState.READY` an acceptable usage?

[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/mongo_client.py]: Answering my own question, no, it is not

File: pymongo/mongo_client.py [hunk 3]
```diff
@@ -2084,16 +2086,16 @@ def __init__(self, client, server, session):
...
+        self.sock_info = None
         self.handled = False
 
     def contribute_socket(self, sock_info):
         """Provide socket information to the error handler."""
         self.max_wire_version = sock_info.max_wire_version
         self.sock_generation = sock_info.generation
         self.service_id = sock_info.service_id
-        self.completed_handshake = True
+        self.sock_info = sock_info
```
Comment (ShaneHarvey) [inline pymongo/mongo_client.py:2098]: contribute_socket is only ever called with an established connection (ie when `sock_info.completed_handshake == True`). So is this change needed at all?

File: pymongo/monitoring.py [hunk 4]
```diff
@@ -1774,7 +1774,7 @@ def publish_connection_check_out_failed(self, address, reason):
         event = ConnectionCheckOutFailedEvent(address, reason)
         for subscriber in self.__cmap_listeners:
             try:
-                subscriber.connection_check_out_started(event)
+                subscriber.connection_check_out_failed(event)
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/monitoring.py:1777]: Can you add a test for this in test_cmap?

File: pymongo/mongo_client.py (hunk 3 above)
Comment (blink1073) [reply pymongo/mongo_client.py:2098]: Oh interesting, let me try that

Comment (blink1073) [reply pymongo/mongo_client.py]: Yeah that works

File: pymongo/monitoring.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/monitoring.py:1777]: Done

File: test/utils.py [hunk 5]
```diff
@@ -79,37 +79,58 @@ def wait_for_event(self, event, count):
...
+        self.func_calls.append(func.__name__)
+        func(self, *args)
+
+    return inner
+
+
 class CMAPListener(BaseListener, monitoring.ConnectionPoolListener):
+    func_calls = []
+
+    @log_func_call
```
Comment (ShaneHarvey) [inline test/utils.py:93]: What do you think about asserting the proper arg type in the method itself (rather than after the fact):
```python
    def connection_created(self, event):
        assert isinstance(event, ConnectionCreatedEvent)
        self.add_event(event)
```

Note that the tests might be failing due to a difference in behavior of load balanced vs non-load balanced pools. Or it could be another bug.

**Comment by blink1073** [comment]
Remaining failure will be fixed as part of https://jira.mongodb.org/browse/PYTHON-2933

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -56,7 +56,7 @@ repos:
   rev: 0.11.1
   hooks:
     - id: doc8
-      args: [--max-line-length=200]
+      args: ["--ignore=D001"]
```
Comment (ShaneHarvey) [inline .pre-commit-config.yaml:59]: Please add a comment describing D001.

File: pymongo/pool.py [hunk 7]
```diff
@@ -1408,6 +1411,8 @@ def get_socket(self, handler=None):
             listeners.publish_connection_check_out_started(self.address)
 
         sock_info = self._get_socket()
+        if handler:
+            handler.contribute_socket(sock_info)
```
Comment (ShaneHarvey) [inline pymongo/pool.py:1415]: Can we revert this part?

File: pymongo/pool.py [hunk 8]
```diff
@@ -539,6 +540,7 @@ def __init__(self, sock, pool, address, id):
         # Support for mechanism negotiation on the initial handshake.
         self.negotiated_mechs = None
         self.auth_ctx = None
+        self.completed_handshake = False
```
Comment (ShaneHarvey) [inline pymongo/pool.py:543]: This can be removed.

File: pymongo/mongo_client.py [hunk 9]
```diff
@@ -2094,8 +2096,8 @@ def __init__(self, client, server, session):
         # completes then the error's generation number is the generation
         # of the pool at the time the connection attempt was started."
         self.sock_generation = server.pool.gen.get_overall()
-        self.completed_handshake = False
         self.service_id = None
+        self.completed_handshake = False
```
Comment (ShaneHarvey) [inline pymongo/mongo_client.py:2100]: Let's revert this change.

File: pymongo/mongo_client.py [hunk 10]
```diff
@@ -2120,7 +2122,6 @@ def handle(self, exc_type, exc_val):
                     "RetryableWriteError"
                 ):
                     self.session._unpin()
-
```
Comment (ShaneHarvey) [inline pymongo/mongo_client.py:2123]: Please revert.

File: pymongo/mongo_client.py [hunk 11]
```diff
@@ -2075,11 +2077,11 @@ class _MongoClientErrorHandler(object):
 
     __slots__ = (
         "client",
+        "completed_handshake",
         "server_address",
         "session",
         "max_wire_version",
         "sock_generation",
-        "completed_handshake",
```
Comment (ShaneHarvey) [inline pymongo/mongo_client.py:2082]: Please revert.

File: pymongo/mongo_client.py [hunk 12]
```diff
@@ -1191,7 +1193,6 @@ def _get_socket(self, server, session):
                     SERVER_TYPE.LoadBalancer,
                 ):
                     session._pin(server, sock_info)
-                err_handler.contribute_socket(sock_info)
```
Comment (ShaneHarvey) [inline pymongo/mongo_client.py:1194]: Revert.

File: pymongo/mongo_client.py [hunk 13]
```diff
@@ -1282,7 +1283,6 @@ def _run_operation(self, operation, unpack_res, address=None):
             server = self._select_server(
                 operation.read_preference, operation.session, address=address
             )
-
```
Comment (ShaneHarvey) [inline pymongo/mongo_client.py:1285]: Revert.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 6 above)
Comment (blink1073) [reply .pre-commit-config.yaml:59]: Done

File: pymongo/pool.py (hunk 7 above)
Comment (blink1073) [reply pymongo/pool.py:1415]: Yep

File: pymongo/pool.py (hunk 8 above)
Comment (blink1073) [reply pymongo/pool.py:543]: Done

File: pymongo/mongo_client.py (hunk 12 above)
Comment (blink1073) [reply pymongo/mongo_client.py:1194]: Done

File: pymongo/mongo_client.py (hunk 13 above)
Comment (blink1073) [reply pymongo/mongo_client.py:1285]: Done

File: pymongo/mongo_client.py (hunk 11 above)
Comment (blink1073) [reply pymongo/mongo_client.py:2082]: Done

File: pymongo/mongo_client.py (hunk 9 above)
Comment (blink1073) [reply pymongo/mongo_client.py:2100]: Done

File: pymongo/mongo_client.py (hunk 10 above)
Comment (blink1073) [reply pymongo/mongo_client.py:2123]: Done

**Review by ShaneHarvey — APPROVED** [review]
great stuff!
